### ISSUE TITLE
Unable to build custom ports

### ISSUE THREAD

**Issue by andrewhotlab** [jrmarino/synth#94]
Hi. First of all, thanks for this great piece of software, which make greatly simpler to maintain our corporate's pkgng repository.
We found this problem when trying to work with our custom ports:
> \# synth build sysutils/myport
> Scan of sysutils/myport failed, it will not be considered.
> There are no valid ports to build.  (Synth must exit)

The same `sysutils/myport` works fine and it passes all checks we are able to make, for example:
> \# portlint /usr/ports/sysutils/myport
> looks fine.

We have also tried after adding as an additional SUBDIR in the file `/usr/ports/sysutils/Makefile`, but synth still fails to build it, with that generic error "scan failed". Can you suggest what might be wrong with our setup?

TIA

**Comment by jrmarino** [comment]
yes, I know what's wrong.
On FreeBSD ports (but not DragonFly DPorts), all *active* ports have to be listed in the category Makefile.  So in you case, you're missing "SUBDIR+= myport" in /usr/ports/sysutils/Makefile  (or something like that). 
This is because freebsd ports hooks (and unhooks) ports by adding or removing them from the category makefile.

**Comment by andrewhotlab** [comment]
Thanks for your reply but, as I wrote, adding that line into `/usr/ports/sysutils/Makefile` does not solve the problem, even if the output changes. Here it is:
>\# grep myport /usr/ports/sysutils/Makefile
>    SUBDIR += myport
>
>\# synth build sysutils/myport
>Encountered issue with sysutils/myport or its dependencies
>  => sysutils/myport (return code = 1)
>Scan of sysutils/myport failed, it will not be considered.
>There are no valid ports to build.  (Synth must exit)

In this case, `sysutils/myports` depends from another regular port (`sysutils/sg3_utils`, which is built without problems by issuing `synth build sysutils/sg3_utils`).

**Comment by jrmarino** [comment]
There's a problem with the makefile of the port itself.
It may reveal itself if you cd to the myport and type "make -V PORTNAME".

**Comment by andrewhotlab** [comment]
Here is the transcript:

> \# cd /usr/ports/sysutils/myport/
> \# make -V PORTNAME
> myport

I can't figure out what can be wrong with the port: I never had problem to build or update the port by hand or by using `port-mgmt/portmaster`. And even `ports-mgmt/portlint` does not reveal any issue.

**Comment by jrmarino** [comment]
So this eliminates a syntax error in the Makefile.
Now you have to start considering content.
For example, is there a listed dependency that doesn't exist in the ports tree?
e.g. LIB_DEPENDS+= mylib.so:devel/myport2  where myport2 isn't listed in devel category Makefile
or maybe there's an ".include" directive that synth can't access.

You could also try something like:
```
env ENTERAFTER=patch synth test editors/nano
```
and when you get the prompt:
```
cd /xports/sysutils/myport && make -V PORTNAME
```
and see if you get a makefile error then

**Comment by andrewhotlab** [comment]
OK, here is the transcript:
````
# env ENTERAFTER=patch synth test editors/nano
After inspection, it has been determined that there are no packages that
require rebuilding; the task is therefore complete.
Starting interactive build of editors/nano
Stand by, building up to the point requested ...
Entering interactive test mode at the builder root directory.
Type 'exit' when done exploring.
# cd /xports/sysutils/myport && make -V PORTNAME
/xports/sysutils/myport: No such file or directory.
````
I guess that's because `/usr/ports/sysutils/myport` is actually a symlink to `/root/ports/sysutils/myport` , which is the working directory checked out from the Subversion repository where we maintain our ports. In fact if I copy the port's directory under` /usr/ports/sysutils`, synth is finally able to finish the build.

Thank very much for your help in finding the cause of this issue!!

Now I have to find a better way to deal with custom ports, I mean: I cannot manually insert and maintain changes in `Makefile` under each category subdir. And I wish to keep our custom ports directories separated from the official Ports directory hierarchy. For the latter, do you think that disabling tmpfs might solve the problem?

**Comment by jrmarino** [comment]
The issue is that symlinks aren't followed under a null mount.
The best way to achieve separation is to create your own category, and define that category as a valid one (see bsd.port.mk I think).  Then you put your custom ports all in the same directory.  You could rsync your category to a tree that synth is configured for (there are several other approaches that work without symlinks as well).

bottom line: you need an external process to merge the stock tree with your custom overlay to produce a new tree.  So we're talking about shell scripts, etc.  It's not too hard.

**Comment by jrmarino** [comment]
FYI, I'm about ready to launch a new ports project with a major objective at supporting enterprises like yours (e.g. outsource ports management including "custom" ports).  It's not a bolt-on to freebsd ports, but an entirely new (and technically superior) project.

The source for the projects are here, under "jrmarino" project.  The new website will be coming any day now (I'm writing the copy as we speak).

**Comment by jrmarino** [comment]
I mention it because it accepts custom ports as standard feature (although this needs additional testing).
Current catalog: http://ravenports.ironwolf.systems/catalog/

**Comment by andrewhotlab** [comment]
I'll explore these methods to deal with custom ports. Thanks again for your suggestions!

Great to know about [Ravenports](/jrmarino/Ravenports)! At this time we use FreeBSD almost exclusively for servers (both ours and of customers of ours), thus that project seems overwhelming for us, but we see that there is a lot of software that still lack good ports for each UNIX-like platform (for that reason we are maintaining in-house some custom ports for FreeBSD): it surely would be great if there could be a unique package system to "rule them all"! :)

We'll keep an eye on Ravenports for sure. Thanks!

**Comment by jrmarino** [comment]
Well, it seems to be a complaint common among commercial users that the wildly untested "bleeding edge" aspect of FreeBSD ports is a major issue (especially when a FreeBSD release goes EOL with immediate ports support terminated) and the quarterly branches aren't consistent nor of good enough quality to be seriously used.  So Ravenports is aimed squarely at FreeBSD-exclusive outfits (as well as those with mixed OS that want the same software suite exactly regardless of OS).   Just a point that I think is worth making.
